### ISSUE TITLE
systemd warns about `unmerged-usr` on NixOS

### ISSUE THREAD

**Issue by ncfavier** [NixOS/nixpkgs#184826]
`systemctl status` started showing `Tainted: unmerged-usr` on NixOS (https://github.com/systemd/systemd/pull/22984), because it thinks NixOS has an unmerged `/usr`, but this concept doesn't apply to NixOS (everything is merged under `/nix/store`, we only keep `/usr/bin/env` and `/bin/sh` for compatibility).

There seem to be plans for "removing support for unmerged-usr" in systemd, so this should probably be addressed, although I'm not sure to what extent this "support" affects NixOS.

Possible solutions:

1. change the NixOS activation script to put both `sh` and `env` in `/usr/bin` and then symlink `/bin` to `/usr/bin`. This is quick and relatively easy to do (migrating existing setups might be a problem), but introduces more "impure" paths than strictly necessary (`/bin/env` and `/usr/bin/sh`).
2. patch systemd in nixpkgs (or just in NixOS?) to remove that flag/check/concept.
3. patch systemd upstream to detect such situations (maybe simply based on os-release?) and not complain.

I'll create an issue at systemd as well.

Pinging people I saw talk about this: @K900 @alyssais @lheckemann

**Comment by trofi** [comment]
Patching it out might be simplest: https://github.com/systemd/systemd/blob/729d2df8065ac90ac606e1fff91dc2d588b2795d/src/core/manager.c#L4503-L4505

**Comment by K900** [comment]
I don't think we _actually_ care, the taint flag has been there for a while, they're just making it louder. Maybe we should add our own `nixos` taint flag? Either way, I'd just go upstream with this and see what they think - after all, the taint flags are there for their convenience.

**Comment by ncfavier** [comment]
Opened https://github.com/systemd/systemd/issues/24191

**Comment by ncfavier** [comment]
Since I agree with Lennart I implemented solution 1: https://github.com/NixOS/nixpkgs/pull/185040

**Comment by NickCao** [comment]
Introducing more global state is likely to be a footgun in the future, I'm against this solution.

**Comment by SuperSandro2000** [comment]
> There seem to be plans for "removing support for unmerged-usr" in systemd, so this should probably be addressed, although I'm not sure to what extent this "support" affects NixOS.

I didn't check it but I would assume that nothing for us changes.

**Comment by winterqt** [comment]
> I don't think we _actually_ care, the taint flag has been there for a while, they're just making it louder.

@K900 How are they making it louder other than the initial addition of the flag?

> I didn't check it but I would assume that nothing for us changes.

@SuperSandro2000 The thing is, I can't find anything on the systemd side as to what exactly "dropping support" for this in systemd will mean. Like, will systemd things start breaking? Or will the maintainers just stop officially supporting our setup?

**Comment by K900** [comment]
> @K900 How are they making it louder other than the initial addition of the flag?

It started showing the taint in `systemctl status`, not just at build time.

**Comment by rnhmjoj** [comment]
Why don't we admit that `/bin/sh` and `/usr/bin/env` are just two special shebangs?
You would never call /bin/sh by hand and if a software does it's probably already completely broken on NixOS.

Let's get rid of these impurities once and for all and just add two binfmt registrations for shell scripts.

**Comment by NickCao** [comment]
Maybe we can have an option to opt out of these two paths for now. `/usr/bin/env` is an easy one, while `/bin/sh` is required by glibc thus may need some workarounds.

**Comment by flokli** [comment]
Having neither /bin nor /usr/bin would definitely have an appeal.

It needs further testing though:

While we should be able to match on /bin/sh and /usr/bin/env, we probably would need to take care of parsing the rest of the line, to handle command arguments, no?

Can we guarantee the binfmt registrations happen early enough, before any scripts with shebangs are executed?

**Comment by NickCao** [comment]
binfmt does not sound like the correct way to go forward, it just replaces one global state with another more obscure and error prone one.

**Comment by rnhmjoj** [comment]
> Maybe we can have an option to opt out of these two paths for now

There are, but are hidden: `environment.usrbinenv` and `environment.binsh`. The former can change and also disable the impurity, the latter can't.

> /bin/sh is required by glibc thus may need some workarounds.

I would patch libc to use dash or runtimeShell. I don't know why this has [never](https://github.com/NixOS/nixpkgs/issues/1424) been done, probably there are reasons.

**Comment by NickCao** [comment]
patching libc would create a cyclic dependency between `glibc` and `sh`

**Comment by rnhmjoj** [comment]
There are surely ways around that. I can think of building sh with musl instead of glibc or combining the libc and sh in a libc-with-sh derivation.

**Comment by flokli** [comment]
This has been addressed in https://github.com/NixOS/nixpkgs/pull/194331, this can be closed.

**Comment by typetetris** [comment]
I know, this is done for now. But what was the problem with solution 1? Why are some additional links impure?
